393 endpoint cache response (#475)
api3dao/vitepress-docs: docs/reference/airnode/latest/understand/configuring.md

   caching enabled it will fulfill the response using the cached API response
   value.

Clearance of the local filesystem cache is dependant on the environment Airnode
runs in. As a guideline:

- AWS Lambda persists for 2.5 hours.
- Google Cloud persistence is completely arbitrary.
- Docker persists as long as the container remains running.

If the local filesystem is cleared, the cache is lost and if a repeat request is
received after the cache has been cleared, Airnode will repeat the API call.
Airnode self-clears cached data after the data has reached 1 hour's age.

Furthermore, as an example: if an Airnode Lambda environment/container has been
"warm" for 2.4 hours and Airnode gets a request, that request will only be
cached until the container resets, in this case 2.5 hours - 2.4 hours = 0.1
hours.

As described above, caching on serverless infrastructure is unreliable and if
request caching is mission critical, Airnode should be run in a Docker container
with a persistent `/tmp` directory.

Caching is useful for non-idempotent API operations like random number
generators. Consider the following use-case:

> > Random Numbers: a malicious blockchain provider can selectively block

Reviewer: To tie the example to caching, perhaps start with "Random Numbers: in the absence of caching, a malicious..."

and that these endpoints can be triggers for `rrp`, `http`, and/or
`httpSignedData` as desired.

#### Considerations: cached responses

> When setting `triggers.rrp[n].cacheResponses = true` Airnode has the ability to
> cache API responses, per request ID, to the local filesystem. This only applies

Reviewer: There isn't any equal signs used in the config, also remove "has the ability" given you're setting to true, so I suggest:
```suggestion
When setting `triggers.rrp[n].cacheResponses` to `true` Airnode caches API responses
on the local filesystem by request ID. This only applies
```

to blockchain requests. The flow is:

1. A request is received from the Blockchain.

> 1. The Airnode does an HTTP/S call to a remote API based on the request.

Reviewer: ```suggestion
1. The Airnode makes an API call based on the request.
```

> 1. [with caching enabled] Airnode stores the response locally, in the local
>    filesystem, per request ID.

Reviewer: "local" twice
```suggestion
1. [with caching enabled] Airnode stores the response on the local
   filesystem by request ID.
```

1. Airnode then follows the usual process of extracting the response,
   post-processing it and sending the response back on chain as a fulfillment
   callback.
1. After all current requests are processed, Airnode then exits until the next
   cycle (the next minute).
1. If during the next cycle, if the Airnode gets the same request and it isn't
   fulfilled, it will try and fulfill the request again. Ordinarily, it would
   follow steps 1, 2 and 4 (as shown above) without caching enabled, but with

>    caching enabled it will read the previous API response value from the local
>    filesystem and send that back to the chain.

Reviewer: ```suggestion
   caching enabled it will fulfill the response using the cached API response value.
```

Clearance of the local filesystem cache is dependant on the environment Airnode
runs in. As a guideline:

- AWS Lambda persists for 2.5 hours.
- Google Cloud persistence is completely arbitrary.
- Docker persists as long as the container remains running.

If the local filesystem is cleared, the cache is lost and if a repeat request is

> received after the cache has been cleared, Airnode will contact the remote API
> again. Airnode self-clears cached data after the data has reached 1 hour's age.

Reviewer: ```suggestion
received after the cache has been cleared, Airnode will repeat the API call.
Airnode self-clears cached data after the data has reached 1 hour's age.
```

Furthermore, as an example: if an Airnode Lambda environment/container has been

> "warm" for 2.4 hours and AIrnode gets a request, that request will only be

Reviewer: Airnode capitalization

cached until the container resets, in this case 2.5 hours - 2.4 hours = 0.1

> hours remaining to the environment being cleared.

Reviewer: ```suggestion
hours.
```

> Caching on serverless infrastructure is unreliable and best effort. If request
> caching is mission critical Airnode should be run in a Docker container with a
> persistent `/tmp` directory.

Reviewer: ```suggestion
As described above, caching on serverless infrastructure is unreliable and if request
caching is mission critical, Airnode should be run in a Docker container with a
persistent `/tmp` directory.
```

Caching is useful for non-idempotent API operations like random number
generators. Consider these two use-cases:

- Consider an endpoint called `https://acme.cool/toggle` that simply toggles
  between two values - you probably wouldn't want something to be toggled
  multiple times for one request (repeated fulfillment attempts over multiple

>   cycles).

Reviewer: I don't understand this example. Perhaps just have the random number example, I think it makes the point clearly.

Author: If a callback fulfillment fails when the toggle is meant to be true, then without caching the retry will make it false. Emanuel wrote this one and I admit it took me a while to "get it" but it is relevant. However I do not think we should make the reader think that hard so I will remove it.

> - Random Numbers; a malicious blockchain provider can selectively block

Reviewer: ```suggestion
- Random Numbers: a malicious blockchain provider can selectively block
```
